### PR TITLE
Elide object safety errors on non-existent trait function

### LINKED ISSUE #58734: Calling nonexistent function on a trait causes spurious object safety errors
```rust
trait Trait {
    fn exists(self) -> ();
    
    fn not_object_safe() -> Self;
}

impl Trait for () {
    fn exists(self) -> () {
    }
    
    fn not_object_safe() -> Self {
        ()
    }
}

fn main() {
    // object-safe or not, this call is OK
    Trait::exists(());
    // but, in addition to nonexistence, this causes an object safety error
    Trait::nonexistent(());
}
```
Gives
```
error[E0599]: no function or associated item named `nonexistent` found for type `dyn Trait` in the current scope
  --> src/main.rs:20:5
   |
20 |     Trait::nonexistent(());
   |     ^^^^^^^^^^^^^^^^^^ function or associated item not found in `dyn Trait`

error[E0038]: the trait `Trait` cannot be made into an object
  --> src/main.rs:20:5
   |
20 |     Trait::nonexistent(());
   |     ^^^^^^^^^^^^^^^^^^ the trait `Trait` cannot be made into an object
   |
   = note: method `not_object_safe` has no receiver

error: aborting due to 2 previous errors
```

As the first call shows, it is in fact possible to call a non-object-safe trait's methods in this manner, but calling a nonexistent method gives an object safety error, adding noise to the output.

Note as well the E0599 error says the receiver type is `dyn Trait`.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f5056d66d32e8c598dbda20783b0169)

### PR DESCRIPTION
**Pull request by estebank**
Fix #58734.

r? @zackmdavis

### CODE REVIEW

**Review by zackmdavis — APPROVED** [review]
I read the diff, but I don't quite understand _why_, fundamentally, we were outputting an object safety error for nonexistent methods? (It would be more elegant to just not do that in the first place, rather than save the spans of nonexistent methods and then use that to decide to not emit the object safety error at the last moment.) But if you can't see any other way to do this, then, yes, r=me

**Comment by estebank** [comment]
> It would be more elegant to just not do that in the first place, rather than save the spans of nonexistent methods and then use that to decide to not emit the object safety error at the last moment. But if you can't see any other way to do this

I agree, and indeed I couldn't find a more elegant way of not emitting the error in the first place. I feel this is a bit of a hack, but better than doing nothing.

@bors r=zackmdavis

File: src/librustc/session/mod.rs [hunk 1]
```diff
@@ -164,6 +164,9 @@ pub struct Session {
 
     /// Cap lint level specified by a driver specifically.
     pub driver_lint_caps: FxHashMap<lint::LintId, lint::Level>,
+
+    /// `Span`s of trait methods that weren't found to avoid emitting object safety errors
+    pub trait_methods_not_found: OneThread<RefCell<FxHashSet<Span>>>,
```
Comment (Zoxc) [inline src/librustc/session/mod.rs]: This is potentially used on multiple threads.

**Pull request merged**